src/hotspot/share/code/nmethod.cpp

```diff
@@ -474,13 +490,30 @@ nmethod* nmethod::new_native_nmethod(const methodHandle& method,
     if (exception_handler != -1) {
       offsets.set_value(CodeOffsets::Exceptions, exception_handler);
     }
-    nm = new (native_nmethod_size, CompLevel_none)
+
+    // MH intrinsics are dispatch stubs which are compatible with NonNMethod space.
+    bool allow_NonNMethod_space = method->is_method_handle_intrinsic();
+    nm = new (native_nmethod_size, allow_NonNMethod_space)
     nmethod(method(), compiler_none, native_nmethod_size,
             compile_id, &offsets,
             code_buffer, frame_size,
             basic_lock_owner_sp_offset,
             basic_lock_sp_offset,
             oop_maps);
+#ifdef ASSERT
+    // Allocating in NonNMethod space is only for special nmethods which don't
+    // need to be findable by nmethod iterators.
+    // GC may not look for Oops, there.
+    if (allow_NonNMethod_space) {
+      assert(oop_maps == nullptr, "expectation");
+      CheckForOopsClosure cfo;
+      nm->oops_do(&cfo);
+      assert(!cfo.found_oop(), "no oops allowed");
+      CheckForMetadataClosure cfm(/* ignore reference to own Method */ nm->method());
```

Comment:
If we are going to ignore method(), then how about checking that it's classloader is parmanent (ClassLoaderData::is_permanent_class_loader_data)?  We also might need a review by a GC expert.  @fisk, do you agree it is safe to put nmethod in NonNMethod space given the above checks?

Reply:
Hmm. The GC used to walk all nmethods an ensure the is_unloading_state is computed for all of them. When using a STW collector, we will crash if trying to compute this outside of the UnloadingScope, as we then don't have rnough information to be able to compute it. So I suppose this is a bit fragile; if an nmethod is acquired from outside of an nmethod iterator, and the is_unloading() question is asked, we risk crashing.
I suppose we could strengthen this by letting is_unloading() know rhat the answer is false for method handle intrinsics.
It feels more and more like method handle intrinsics shouldn't be nmethods as more and more code that deals with nmethods needs to know that this isn't really an nmethod. Although that's not a new observation for this patch.

Reply:
So, calling `IsUnloadingBehaviour::is_unloading` is unsafe. Should we only call it when `is_permanent_class_loader_data()` for the method returns false?

Reply:
Does this also mean that nmethods for method handle intrinsics could previously get unloaded when they became "cold", and now if they are in NonNMethod they live forever?  Someone could probably write a test that fills up NonNMethod with method handle intrinsics for various signatures by spinning up temporary classes with random signatures.

Reply:
We were able to hit the `IsUnloadingBehaviour::is_unloading` problem! I'm testing a workaround, now.
Note that the Methods are also permanently kept in `_invoke_method_intrinsic_table`. Getting rid of method handle adapters is currently not supported.